### PR TITLE
Time and weather implementation

### PR DESCRIPTION
**Pull request by AsgardXIV**
Implements time and weather changing.

`WorldHooks` in this case is actually doing a bit more than hooking and the general vars it exposes might be better suited to some other more generic 'Service' but there doesn't seem to be a model for that in Ktisis currently, unless I missed it.

Also not crazy about how it's grabbing the territory/weather info from Lumina (The animation stuff has a similar issue), so that's another area where an official way to do that kind of work might be nice in future.

### CODE REVIEW

File: Ktisis/Interface/Components/WeatherControls.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+				var weather = WeatherSheet.Value.SingleOrDefault(i => i.RowId == wr.Weather);
+
+				if (weather == null)
+					continue;
+
+				if (ZoneValidWeatherList.Count(x => x.Id == (byte)weather.RowId) == 0)
+					ZoneValidWeatherList.Add(((byte)weather.RowId, weather.Name));
+
+				ZoneValidWeatherList.Sort((x, y) => x.Id.CompareTo(y.Id));
```
Comment (Fayti1703) [inline Ktisis/Interface/Components/WeatherControls.cs:97]: Sorting the list `n` times seems unnecessary here -- move this `.Sort` call after the loop instead.

File: Ktisis/Interop/Hooks/WorldHooks.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+			var etAddress = Services.SigScanner.ScanText("48 89 5C 24 ?? 57 48 83 EC ?? 48 8B F9 48 8B DA 48 81 C1 ?? ?? ?? ?? E8 ?? ?? ?? ?? 4C 8B 87 ?? ?? ?? ??");
+			UpdateEorzeaTimeHook = Hook<UpdateEorzeaTimeDelegate>.FromAddress(etAddress, UpdateEorzeaTime);
+			UpdateEorzeaTimeHook.Enable();
+
+			var twAddress = Services.SigScanner.ScanText("48 89 5C 24 ?? 55 56 57 48 83 EC ?? 48 8B F9 48 8D 0D ?? ?? ?? ??");
+			UpdateTerritoryWeatherHook = Hook<UpdateTerritoryWeatherDelegate>.FromAddress(twAddress, UpdateTerritoryWeather);
+			UpdateTerritoryWeatherHook.Enable();
+
+			IntPtr rawWeather = Services.SigScanner.GetStaticAddressFromSig("4C 8B 05 ?? ?? ?? ?? 41 8B 80 ?? ?? ?? ?? C1 E8 02");
+			WeatherSystem = (WeatherSystem*)(*(IntPtr*)rawWeather);
```
Comment (Fayti1703) [inline Ktisis/Interop/Hooks/WorldHooks.cs:47]: You can simplify this expression by casting to `WeatherSystem**` instead:
```suggestion
			WeatherSystem = *(WeatherSystem**) rawWeather;
```

**Review by Fayti1703 — CHANGES_REQUESTED** [review]
Looks good overall; there's just two things I'd like to see changed here, see below.

It might also be worth contributing the fields in `ExtendedFramework` upstream to `FFXIVClientStructs` (though this isn't a requirement for merging).

**Comment by Fayti1703** [comment]
Relations:
* Conflicts #87's bi-state enum removal [severity: minor, ordering: arbitrary]
* Conflicts #89 [severity: minor, ordering: arbitrary]

*(This is for internal tracking. Conflicts need not be resolved prior to merging.)*

File: Ktisis/Interface/Components/WeatherControls.cs (hunk 1 above)
Comment (AsgardXIV) [reply Ktisis/Interface/Components/WeatherControls.cs:97]: Oh yeh, that's not intentional. I think that's a refactor/copy-paste mistake when I was cleaning up.

**Comment by AsgardXIV** [comment]
Fixes are up, and I do intend to push the changes back to ClientStructs so we can drop ExtendedFramework once that makes it back upstream.

**Comment by AsgardXIV** [comment]
Update for 6.3

**Comment by Fayti1703** [comment]
It looks like you merged your old and new histories rather than force-pushing, which is making the commit history look a bit wierd.
I could fix that if you want, but I'd rather not force-push someone else's branch without their permission.

**Comment by AsgardXIV** [comment]
Should be resolved

**Comment by chirpxiv** [comment]
Confused why these (#96, #97, #98) got closed, is something wrong?

**Comment by AsgardXIV** [comment]
Yes, as per my twitter and posts on discord, I am taking a break from the FFXIV dev community and won't be actively contributing to any projects at present. The code is still there in the PR if you want to use it, I just won't be shepherding it through the process. 

I'm happy to discuss it privately if you have any questions, a GitHub issue is not the right forum.

**Comment by chirpxiv** [comment]
Okay. These can stay open though, we'll make any adjustments as needed.

**Comment by chirpxiv** [comment]
> I'm happy to discuss it privately if you have any questions, a GitHub issue is not the right forum.

Side note regarding this - I'd prefer that any communication regarding Ktisis stays public for the sake of transparency, for users, contributors, and the rest of the team. If you don't want to discuss it in a GitHub issue, then the Discord has the #ktisis-dev channel for that.

**Comment by AsgardXIV** [comment]
> Side note regarding this - I'd prefer that any communication regarding Ktisis stays public for the sake of transparency, for users, contributors, and the rest of the team. If you don't want to discuss it in a GitHub issue, then the Discord has the #ktisis-dev channel for that.

In the interests of transparency then. 

The very serious accusations of copying and theft, as well as the vile abuse I've received in DMs on both Discord and Twitter have been from people claiming or implying they are acting on behalf of the Ktisis dev team, without being willing to use anything more than a socketpuppet account and immediately lock down the account or delete the tweets/account in question.
I certainly hope that isn't actually the case and that these are trolls or people white knighting for you without your knowledge.

Regardless, it's clear that this has turned into some perverse, childish, toxic competition between Ktisis/Brio/Anamnesis and if that's going to be the case, I don't want to be involved in posing development for the foreseeable future. I did this for the fun of it and in pursuit of knowledge that would benefit all 3 projects, and to be attacked, called homophobic names and abused is not my idea of fun.

**Comment by Fayti1703** [comment]
I'm really sorry to hear that. I completely understand you wanting to take a break if that's what you've had to deal with.

We were genuinely unaware of any harassment. We did not and do not direct others to abuse you or other developers; nor did or will we use alt accounts to hurt other people.
Whoever sent you those messages is not affiliated with us and acted without our consent or knowledge.

Speaking for myself, I realize I may come off as harsh or abrasive while critiquing something on a technical level, but that is not meant as a personal attack in any way.


Regardless, we respect your decision and wish you the best.

**Comment by AsgardXIV** [comment]
> We were genuinely unaware of any harassment. We did not and do not direct others to abuse you or other developers; nor did or will we use alt accounts to hurt other people. Whoever sent you those messages is not affiliated with us and acted without our consent or knowledge.

I'm glad to hear that, and I believed it to be the case, but the confirmation is a comfort.

> Speaking for myself, I realize I may come off as harsh or abrasive while critiquing something on a technical level, but that is not meant as a personal attack in any way.

I have no problems with anything you've said to me personally or our interactions, discussion of code quality is healthy and normal.

> Regardless, we respect your decision and wish you the best.

I wish you the best as well, perhaps in the future when things have calmed down, we'll be able to collab again.

**Pull request closed without merging**